### DISCUSSION TITLE
地图匹配报错

### DISCUSSION THREAD

**Discussion by Murphy666-code** [zdsjjtTLG/TrackIt#20] [Q&A]
运行代码如下：# 1. 从gotrackit导入相关模块Net, MapMatch
import pandas as pd
import geopandas as gpd
from gotrackit.map.Net import Net
from gotrackit.MapMatch import MapMatch

if __name__ == '__main__':

    # 1. Read GPS data
    # This is a file with GPS data of 10 vehicles, which has been cleaned and segmented
    # GPS data for map matching needs to be cleaned and segmented by the user
    gps_df = pd.read_csv(r'D:\数据清洗实验.csv')
    print(gps_df)

    # 2.To build a net, the road network link layer and road network node layer must be in WGS-84, EPSG:4326 geographic coordinate system
    # Please pay attention to the encoding of the shp file. You can specify the encoding to ensure that there is no garbled code in the field.
    link = gpd.read_file(r'D:\python处理\LinkAfterModify.shp')
    node = gpd.read_file(r'D:\python处理\NodeAfterModify.shp')
    my_net = Net(link_gdf=link,
                 node_gdf=node)
    my_net.init_net()  # initialization

    # 3. map-match
    mpm = MapMatch(net=my_net, gps_buffer=100, flag_name='xa_sample',
           use_sub_net=True, use_heading_inf=True, omitted_l=6.0,
           del_dwell=True, dwell_l_length=50.0, dwell_n=0,
           export_html=True, export_geo_res=True, use_gps_source=True,
           export_all_agents=False,
           out_fldr=r'D:\地图匹配实验', dense_gps=False,
           gps_radius=50.0)

    # The first result returned is the matching result table
    # The second is the relevant information of the warning
    # The third is the ID number list of the agent with matching errors
    match_res, may_error_info, error_info = mpm.execute(gps_df=gps_df)
    print(match_res)
    match_res.to_csv(r'./data/output/match_visualization/xa_sample/match_res.csv', encoding='utf_8_sig', index=False)
报错如下：Traceback (most recent call last):
  File "C:\Users\Sha mofei\PycharmProjects\pythonProject\数据读取\地图匹配.py", line 19, in <module>
    my_net = Net(link_gdf=link,
             ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\WrapsFunc.py", line 13, in inner
    res = f(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\map\Net.py", line 123, in __init__
    self.__node = Node(node_gdf=node_gdf, is_check=is_check, init_available_node=self.cache_id,
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\map\Node.py", line 35, in __init__
    self.planar_crs = judge_plain_crs_based_on_node(node_gdf=self.__node_gdf)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\tools\geo_process.py", line 297, in judge_plain_crs_based_on_node
    assert -180.0 <= mean_x <= 180.0, 'the value of lng must in [-180, 180]'
           ^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: the value of lng must in [-180, 180]

Process finished with exit code 1

**Comment by zdsjjtTLG** [comment]
路网坐标系不对，不是EPSG:4326，另外不要看英文文档，很久没维护了，看这个：https://gotrackit.readthedocs.io/en/latest/index.html

**Reply by Murphy666-code** [reply]
哦哦好的 路网就是根据手册代码生成的：import geopandas as gpd
import gotrackit.netreverse.NetGen as ng

if __name__ == '__main__':
    # 对link.shp的要求: 只需要有geometry字段即可, 但是geometry字段的几何对象必须为LineString类型(不允许Z坐标)
    link_gdf = gpd.read_file(r'D:\交通\2020路网李\python处理\2020路网及其句法值.shp')
    print(link_gdf)
    # update_link_field_list是需要更新的路网基本属性字段：link_id，from_node，to_node，length，dir
    # 示例中：link_gdf本身已有dir字段，所以没有指定更新dir
    new_link_gdf, new_node_gdf, node_group_status_gdf \
        = ng.NetReverse.create_node_from_link(link_gdf=link_gdf, using_from_to=False,
                                              update_link_field_list=['link_id',
                                                                      'from_node',
                                                                      'to_node',
                                                                      'length'],
                                              plain_crs='EPSG:32650',
                                              modify_minimum_buffer=0.7,
                                              execute_modify=True,
                                              ignore_merge_rule=True,
                                              out_fldr=r'D:\交通\2020路网李\python处理')
坐标系有问题的话怎末处理

**Reply by Murphy666-code** [reply]
坐标问题我解决了 但匹配出来之后有以下问题：
![微信图片_20241227163225](https://github.com/user-attachments/assets/25d026fe-e350-41ad-85ed-b25f1e756d8b)
我的gps点位数据比较稀疏 有一些断链的地方 还有一些标红路段 这是调参的问题吗

**Reply by Murphy666-code** [reply]
还有一些报错：create_computational_net costs :0.042000532150268555 seconds!
C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\model\Markov.py:309: UserWarning: the GPS point with seq: [5] is not associated with any candidate road segment 
                            and will not be used for path matching calculation...
  warnings.warn(rf'''the GPS point with seq: {_gap} is not associated with any candidate road segment
__generate_st costs :0.10566568374633789 seconds!
C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\model\Markov.py:858: UserWarning: gps seq: 22 -> 25 problem with state transfer
                            from_link:(5905, 2745) -> to_link:(10347, 11261)
  warnings.warn(
C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\model\Markov.py:886: UserWarning: gps seq: 28 -> 29 problem with state transfer
                            from_link:(9984, 476) -> to_link:(11151, 7442)
  warnings.warn(
C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\model\Markov.py:886: UserWarning: gps seq: 34 -> 35 problem with state transfer
                            from_link:(1266, 7782) -> to_link:(7343, 8720)
  warnings.warn(
C:\Users\Sha mofei\AppData\Roaming\Python\Python312\site-packages\gotrackit\model\Markov.py:858: UserWarning: gps seq: 44 -> 48 problem with state transfer
                            from_link:(11002, 1535) -> to_link:(6161, 6908)
  warnings.warn(

**Reply by zdsjjtTLG** [reply]
这些都是WARNING，表明你的路网可能存在缺失.路径匹配本质是把定位点匹配到你自己的路网上，如果GPS点在你的路网上走不通，那就会报警告，当然也有可能是你的参数不合理导致正确的路段没有进入你的候选范围，文档有讲可能的原因，好好看看文档吧

**Reply by zdsjjtTLG** [reply]
你这么想，按照GPS的走向，你觉得他应该走哪条路？但是实际他没这么走，是不是可能因为路网本身就缺失了部分路段？

**Reply by zdsjjtTLG** [reply]
关注路网是否缺失、路段允许的行车方向是否正确、匹配参数是否合理，文档都有讲的

**Reply by zdsjjtTLG** [reply]
该修路网修路网，该调整参数就调整参数，看看快速开始里面的案例吧
